File: LeetCode/Java/searchBST.java
/*
    700. Search in a Binary Search Tree
    https://leetcode.com/problems/search-in-a-binary-search-tree/
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

public TreeNode searchBST(TreeNode root, int val) {
    if (root == null)
        return null;
    if (root.val == val)
        return root;
    else if (root.val > val)
        return searchBST(root.left, val);
    else 
        return searchBST(root.right, val);
}

/*
    Runtime: 0 ms, faster than 100.00% of Java online submissions for Search in a Binary Search Tree.
    Memory Usage: 38.4 MB, less than 100.00% of Java online submissions for Search in a Binary Search Tree.
 */

File: LeetCode/Swift/searchBST.swift
/*
    700. Search in a Binary Search Tree
    https://leetcode.com/problems/search-in-a-binary-search-tree/
 */

class Solution {
    func searchBST(_ root: TreeNode?, _ val: Int) -> TreeNode? {
        if (root == nil) {
            return nil;
        }
        if (root!.val == val) {
            return root;
        } else if (root!.val > val) {
            return searchBST(root!.left, val);
        } else {
            return searchBST(root!.right, val);
        }
    }
}

/*
    Runtime: 176 ms, faster than 66.89% of Swift online submissions for Search in a Binary Search Tree.
    Memory Usage: 21.3 MB, less than 100.00% of Swift online submissions for Search in a Binary Search Tree.
 */
